src/main/java/software/amazon/nio/spi/s3/S3ClientProvider.java

```diff
@@ -0,0 +1,288 @@
+     * This is the same as universalClient(false);
+     *
+     * @return a S3Client not bound to a region
+     */
+    public S3Client universalClient() {
+        return universalClient(false);
+    }
+
+    /**
+     * This method returns a universal client (i.e.not bound to any region)
+     * that can be used by certain S3 operations for discovery
+     *
+     * @param async true to return an asynchronous client, false otherwise
+     * @return a S3Client not bound to a region
+     */
+    public <T extends AwsClient> T universalClient(boolean async) {
+        return (T)((async) ? DEFAULT_ASYNC_CLIENT : DEFAULT_CLIENT);
+    }
+
+    /**
+     * Generate a client for the named bucket using a default client to determine the location of the named bucket
+     * @param bucketName the named of the bucket to make the client for
+     * @return an S3 client appropriate for the region of the named bucket
+     */
+    protected S3Client generateClient(String bucketName){
+        //
+        // TODO: use generics like in universalClient()
+        //
+        return generateClient(bucketName, universalClient());
+    }
+
+    /**
+     * Generate an asynchronous client for the named bucket using a default client to determine the location of the named bucket
+     * @param bucketName the named of the bucket to make the client for
+     * @return an asynchronous S3 client appropriate for the region of the named bucket
+     */
+    protected S3AsyncClient generateAsyncClient(String bucketName){
+        return generateAsyncClient(bucketName, DEFAULT_CLIENT);
+    }
+
+    /**
+     *
+     */
+    protected S3Client generateClient (String bucketName, S3Client locationClient) {
+        logger.debug("generating client for bucket: '{}'", bucketName);
+        S3Client bucketSpecificClient = null;
+
+        String bucketLocation = null;
+        try {
+            logger.debug("determining bucket location with getBucketLocation");
+            bucketLocation = locationClient.getBucketLocation(builder -> builder.bucket(bucketName)).locationConstraintAsString();
+            bucketSpecificClient = this.clientForRegion(bucketLocation);
+        } catch (S3Exception e) {
+            if(e.statusCode() == 403) {
+                logger.debug("Cannot determine location of '{}' bucket directly. Attempting to obtain bucket location with headBucket operation", bucketName);
+                try {
+                    final HeadBucketResponse headBucketResponse = locationClient.headBucket(builder -> builder.bucket(bucketName));
+                    bucketSpecificClient = this.clientForRegion(headBucketResponse.
+                            sdkHttpResponse()
+                            .firstMatchingHeader("x-amz-bucket-region")
+                            .orElseThrow(() -> new NoSuchElementException("Head Bucket Response doesn't include the header 'x-amz-bucket-region'")));
+                } catch (S3Exception e2) {
```

Comment:
I do not find it that inelegant, actually. It could probably be refactored with some lambdas to be invoked on errors, but I would keep it as is for now.